### ISSUE TITLE
InternalServerError on BasicAuth protected registry

### ISSUE THREAD

**Issue by marcuslinke** [brennerm/docker-registry-frontend#2]
Using the docker image after configuring registry url I got the following exception:

```
[2017-07-06 12:24:00,923] ERROR in app: Exception on / [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.5/dist-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.5/dist-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.5/dist-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/usr/local/lib/python3.5/dist-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.5/dist-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "main.py", line 54, in registry_overview
    registries=registry_web.registries)
  File "/usr/local/lib/python3.5/dist-packages/flask/templating.py", line 134, in render_template
    context, ctx.app)
  File "/usr/local/lib/python3.5/dist-packages/flask/templating.py", line 116, in _render
    rv = template.render(context)
  File "/usr/local/lib/python3.5/dist-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/local/lib/python3.5/dist-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.5/dist-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/root/templates/registry_overview.html", line 1, in top-level template code
    {% extends "layout.html" %}
  File "/root/templates/layout.html", line 21, in top-level template code
    <div id="content" class="container-fluid">{% block content %}{% endblock %}</div>
  File "/root/templates/registry_overview.html", line 29, in block "content"
    <td>{{ registry.get_number_of_repos() }}</td>
  File "/root/docker_registry_frontend/registry.py", line 79, in get_number_of_repos
    return len(self.get_repos())
  File "/root/docker_registry_frontend/registry.py", line 315, in get_repos
    url=self._url)
  File "/root/docker_registry_frontend/registry.py", line 53, in json_request
    DockerRegistry.string_request(*args, **kwargs)
  File "/root/docker_registry_frontend/cache.py", line 21, in decorator
    result = f(*args, **kwargs)
  File "/root/docker_registry_frontend/registry.py", line 59, in string_request
    return DockerRegistry.request(*args, **kwargs).read().decode()
  File "/root/docker_registry_frontend/registry.py", line 64, in request
    return urllib.request.urlopen(request, timeout=3)
  File "/usr/lib/python3.5/urllib/request.py", line 163, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.5/urllib/request.py", line 472, in open
    response = meth(req, response)
  File "/usr/lib/python3.5/urllib/request.py", line 582, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.5/urllib/request.py", line 510, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 444, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 590, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 401: Unauthorized
127.0.0.1 - - [06/Jul/2017 12:24:00] "GET / HTTP/1.0" 500 
```

**Comment by brennerm** [comment]
Hey @marcuslinke,
Thanks for opening this issue. Support for Basic Auth protected registries is already planned.
I'll close this issue as soon as support has been added.

**Comment by pavlovpetr** [comment]
Hello @brennerm 

I got the next issue when I was trying to get a list of images:

![error](https://user-images.githubusercontent.com/36917400/40712528-9ed7003a-6406-11e8-8097-378666864890.png)

Though I can see the repositories:

![name_of _repos](https://user-images.githubusercontent.com/36917400/40713754-62e636b0-6409-11e8-9d50-4eb9d11b6e6c.png)


"docker logs -f reg-ui" shows following:

```127.0.0.1 - - [30/May/2018 09:32:53] "GET /registry/registry-it/repo/devex36 HTTP/1.0" 500 -
[2018-05-30 09:37:17,166] ERROR in app: Exception on /registry/registry-it/repo/iisserver [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "frontend.py", line 137, in tag_overview
    repo=urldecode_filter(repo))
  File "/usr/local/lib/python3.6/site-packages/flask/templating.py", line 135, in render_template
    context, ctx.app)
  File "/usr/local/lib/python3.6/site-packages/flask/templating.py", line 117, in _render
    rv = template.render(context)
  File "/usr/local/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/local/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/root/templates/tag_overview.html", line 1, in top-level template code
    {% extends "layout.html" %}
  File "/root/templates/layout.html", line 21, in top-level template code
    <div id="content" class="container-fluid">{% block content %}{% endblock %}</div>
  File "/root/templates/tag_overview.html", line 21, in block "content"
    <td>{{ registry.get_size_of_layers(repo, tag) | to_mb }} MB</td>
  File "/root/docker_registry_frontend/registry.py", line 97, in get_size_of_layers
    result += self.get_size_of_layer(repo, image_id)
  File "/root/docker_registry_frontend/cache.py", line 21, in decorator
    result = f(*args, **kwargs)
  File "/root/docker_registry_frontend/registry.py", line 337, in get_size_of_layer
    method='HEAD'
  File "/root/docker_registry_frontend/registry.py", line 64, in request
    return urllib.request.urlopen(request, timeout=3)
  File "/usr/local/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/usr/local/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/local/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/usr/local/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/local/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found
127.0.0.1 - - [30/May/2018 09:37:17] "GET /registry/registry-it/repo/iisserver HTTP/1.0" 500 - 
```

But "curl" returns the folowing:
`curl http://192.168.36.37:5000/v2/iisserver/tags/list`
{"name":"iisserver","tags":["base"]}

Docker version: 18.03.1-ce.

**Comment by mikaellindemann** [comment]
Hi @brennerm,
I can see that some progress are done on this, at least there is a section in the new registry template, as well as creation of registries accepts usernames and passwords.
What is the needed effort to put the last pieces together?

**Comment by brennerm** [comment]
Finally found some time to work on this. Please have a look at #13 and tell me what you think.

**Comment by brennerm** [comment]
Fixed with #13
